packages/server/oauth2/tokenHandler.ts

```diff
@@ -0,0 +1,120 @@
+import getKysely from '../postgres/getKysely'
+import encodeAuthToken from '../utils/encodeAuthToken'
+
+const tokenHandler = uWSAsyncHandler(async (res: HttpResponse, _req: HttpRequest) => {
+  try {
+    const rawBody = await parseBody({
+      res,
+      parser: (buffer: Buffer) => buffer.toString()
+    })
+
+    if (!rawBody) {
+      res.writeStatus('400 Bad Request')
+      res.end(
+        JSON.stringify({error: 'invalid_request', error_description: 'Request body is missing'})
+      )
+      return
+    }
+
+    const params = new URLSearchParams(rawBody)
+    const body: Record<string, string> = {}
+    params.forEach((value, key) => {
+      body[key] = value
+    })
+
+    const {grant_type, code, redirect_uri, client_id, client_secret} = body
+
+    if (grant_type !== 'authorization_code') {
+      res.writeStatus('400 Bad Request')
+      res.end(JSON.stringify({error: 'unsupported_grant_type'}))
+      return
+    }
+
+    if (!code || !redirect_uri || !client_id || !client_secret) {
+      res.writeStatus('400 Bad Request')
+      res.end(JSON.stringify({error: 'invalid_request'}))
+      return
+    }
+
+    // 1. Validate Client Credentials
+    const pg = getKysely()
+    const provider = await pg
+      .selectFrom('OAuthAPIProvider')
+      .selectAll()
+      .where('clientId', '=', client_id)
+      .executeTakeFirst()
+
+    if (!provider || provider.clientSecret !== client_secret) {
+      res.writeStatus('401 Unauthorized')
+      res.end(JSON.stringify({error: 'invalid_client'}))
+      return
+    }
+
+    // 2. Validate Code (atomically delete to prevent replay attacks)
+    const oauthCode = await pg
+      .deleteFrom('OAuthAPICode')
+      .where('id', '=', code)
+      .returning(['id', 'userId', 'expiresAt', sql<string[]>`to_json(scopes)`.as('scopes')])
+      .executeTakeFirst()
+
+    if (!oauthCode) {
+      res.writeStatus('400 Bad Request')
+      res.end(JSON.stringify({error: 'invalid_grant'}))
+      return
+    }
+
+    // 3. Validate Code Expiration
+    const now = new Date()
+    const expiresAt = new Date(oauthCode.expiresAt)
+    if (expiresAt < now) {
+      res.writeStatus('400 Bad Request')
+      res.end(JSON.stringify({error: 'invalid_grant', error_description: 'Code expired'}))
+      return
+    }
+
+    // 4. Validate Redirect URI
+    if (!provider.redirectUris || !provider.redirectUris.includes(redirect_uri)) {
+      res.writeStatus('400 Bad Request')
+      res.end(JSON.stringify({error: 'invalid_grant', error_description: 'Redirect URI mismatch'}))
+      return
+    }
+
+    // 5. Ensure scopes is an array (should always be true with to_json)
+    const scopes = Array.isArray(oauthCode.scopes) ? oauthCode.scopes : []
+
+    // Generate Access Token
+    const userId = oauthCode.userId
+    const authToken = new AuthToken({
+      sub: userId,
+      tms: [],
+      scp: scopes,
+      lifespan_ms: ms('30d')
+    })
+    authToken.iss = 'parabol-oauth2'
```

Comment:
-1 issuer should be the issuing authority. If we don't include the URI, then every PPMI will have the same iss as the cloud-saas, which isn't what we want. I think you meant `aud`, which is the audience, or recipient.